clang/lib/Sema/SemaExpr.cpp

```diff
@@ -9150,6 +9150,15 @@ Sema::CheckAssignmentConstraints(QualType LHSType, ExprResult &RHS,
   LHSType = Context.getCanonicalType(LHSType).getUnqualifiedType();
   RHSType = Context.getCanonicalType(RHSType).getUnqualifiedType();
 
+  // __builtin_counted_by_ref cannot be assigned to a variable, used in
+  // function call, or in a return.
+  if (auto *CE = dyn_cast<CallExpr>(RHS.get())) {
```

Comment:
`RHS.get()->IgnoreParenCasts()`?

Reply:
That's fine I suppose. Though the vast number of `Ignore...Casts` methods and whether they are "good to use" or not is a bit annoying...

Actually, do I need to handle any arbitrary `Expr's` that may have a `__builtin_counted_by_ref` somewhere in it? Maybe something like:

```c
char *c = (char *)(__builtin_counted_by_ref(p->array) + 42);
```

Or is that overkill?

Reply:
Right. As I commented here https://discourse.llvm.org/t/rfc-introducing-new-clang-builtin-builtin-get-counted-by/80836/44?u=gwelymernans, the builtin should be either immediately dereferenced or immediately casted to a boolean type. Otherwise, it's a violation.

That means we should emit an error for builtin being used on binary, unary, and ternary operands (non-condition), except for the above use cases.

Reply:
Done. PTAL.

Reply:
Thanks for the quick iteration! I believe these checks should be good enough for now! I think still there are some cases that should be prevented, e.g.,  
`__builtin_counted_by_ref(p->array)[4] = 0` and `*(__builtin_counted_by_ref(p->array) + 5) = 0`, but I'm fine with filing an issue for now and moving on!

Reply:
Is there a centralized place where I could place a check for both of those?

Reply:
I added some checks for LHS usage. We can probably expand it in the future if need be.